### PR TITLE
spatial flux rework

### PR DESCRIPTION
**Pull request by agrrr3**
discussion:
https://freeorion.org/forum/viewtopic.php?p=104863

missing:
graphics for the composite are missing (and the lance could be better), those are already requested on the forums

https://freeorion.org/forum/viewtopic.php?f=10&t=12175&p=108300
https://freeorion.org/forum/viewtopic.php?f=10&t=12176

### CODE REVIEW

**Comment by agrrr3** [comment]
I am not able to interpret the CI failures... TestPythonParser (Failed) - as far as I can see this PR should not be able to break anything there

**Comment by geoffthemedio** [comment]
> TestPythonParser (Failed) - as far as I can see this PR should not be able to break anything there

I'm not sure, as @o01eg set up those tests, but it looks like the Python parser tests also run the existing C++ / Boost tests on the **.focs.txt** files, and [checks the resulting number of techs](https://github.com/freeorion/freeorion/blob/master/test/parse/TestPythonParser.cpp#L212). Since you're adding more techs, the numbers don't match what it expects.

**Comment by agrrr3** [comment]
> tests on the **.focs.txt** files, and [checks the resulting number of techs](https://github.com/freeorion/freeorion/blob/master/test/parse/TestPythonParser.cpp#L212). Since you're adding more techs, the numbers don't match what it expects.

thanks, missed that part of the test. guess the test is brittle though

**Comment by geoffthemedio** [comment]
I think the purpose is to let him check that the Python and Boost parsers produce the same results, and that specific test is perhaps just there as a sanity check or warning that something was changed so that whatever Python scripts are being parsed might need updating?

**Comment by agrrr3** [comment]
reworked the stealth effect and fixed the lance damage (forgot it does only half the damage as it is a close range weapon)

**Comment by agrrr3** [comment]
playtested this, good to go (after review)

**Comment by agrrr3** [comment]
@o01eg @geoffthemedio can we merge, please

**Comment by geoffthemedio** [comment]
I don't have time to look at this until later this week

File: default/scripting/ship_parts/ShortRange/SR_FLUX_LANCE.focs.txt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    ]
+    mountableSlotTypes = Internal
+    buildcost = 4 * [[FLEET_UPKEEP_MULTIPLICATOR]] * [[SHIP_PART_COST_MULTIPLIER]]
+    buildtime = 2
+    tags = [ "PEDIA_PC_DIRECT_WEAPON" ]
+    location = And [
+        Planet
+        OwnedBy empire = Source.Owner
+    ]
+    icon = "icons/ship_parts/flux-lance.png"
```
Comment (geoffthemedio) [inline default/scripting/ship_parts/ShortRange/SR_FLUX_LANCE.focs.txt:46]: is this file .png missing?

File: default/scripting/ship_hulls/spatial_flux/spatial_flux.macros [hunk 2]
```diff
@@ -0,0 +1,17 @@
+SPATIAL_FLUX_STEALTH_EFFECTS
+'''EffectsGroup
+    scope = Source
+    activation = (Source.ArrivedOnTurn >= CurrentTurn - 1) 
+    accountinglabel = "SPATIAL_FLUX_MALUS"
+    effects = SetStealth value = Value - 30
```
Comment (geoffthemedio) [inline default/scripting/ship_hulls/spatial_flux/spatial_flux.macros]: should make `30` a named value reference and use in the text

File: default/scripting/ship_hulls/spatial_flux/spatial_flux.macros [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+EffectsGroup
+    scope = Source
+    activation = Not Aggressive
+    accountinglabel = "SPATIAL_FLUX_BONUS"
+    effects = SetStealth value = Value + 10 + ( NamedRealLookup name = "SPATIAL_FLUX_STEALTH_TECH_BONUS" *
```
Comment (geoffthemedio) [inline default/scripting/ship_hulls/spatial_flux/spatial_flux.macros]: should make `10` a named reference and use in text

File: default/stringtables/en.txt [hunk 4]
```diff
@@ -15471,12 +15502,27 @@ Spatial Flux Bubble Hull
...
+'''Expensive special purpose built ship built around a single core part using [[tech SHP_SPACE_FLUX_COMPOSITION]]. It has one core slot and one external slot, and its max [[metertype METER_STRUCTURE]] is high.
+
+[[FLUX_STEALTH_DESC]]
 
 [[metertype METER_DETECTION]] is medium and [[metertype METER_SPEED]] is average.
 
 [[BLD_SHIPYARD_BASE_ORBITAL_DRYDOCK_REQUIRED]]'''
 
+FLUX_STEALTH_DESC
+[[metertype METER_STEALTH]] starts at 15, and is increased by 10 when passive. Technology advances such as [[tech SPY_STEALTH_PART_1]] allow crews to increase the stealth of the craft by [[value SPATIAL_FLUX_STEALTH_TECH_BONUS]] each. If the ship stops in a system, the flux coherence breaks down decreasing stealth by 30 for two turns.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]:
```suggestion
Base [[metertype METER_STEALTH]] of ships with this hull is 15 when in a fleet that is set to initiate combat, and 25 for other fleet aggression settings.

Researching each of [[tech SPY_STEALTH_PART_1]], [[tech SPY_STEALTH_PART_2]], [[tech SPY_STEALTH_PART_3]]), and [[tech SPY_STEALTH_4]] will increase base stealth by [[value SPATIAL_FLUX_STEALTH_TECH_BONUS]].

 If the ship stops in a system, its stealth is reduced by 30 for 2 turns, due to breakdown of flux coherence.
```

File: default/stringtables/en.txt [hunk 5]
```diff
@@ -15471,12 +15502,27 @@ Spatial Flux Bubble Hull
...
+
+[[metertype METER_DETECTION]] is medium and [[metertype METER_SPEED]] is average.
+
+[[BLD_SHIPYARD_BASE_ORBITAL_DRYDOCK_REQUIRED]]'''
+
+SH_SPACE_FLUX_COMPOSITE
+Spatial Flux Composite Hull
+
+SH_SPACE_FLUX_COMPOSITE_DESC
+'''Expensive special purpose built ship built around a single core part using [[tech SHP_SPACE_FLUX_COMPOSITION]]. It has one core slot and one external slot, and its max [[metertype METER_STRUCTURE]] is high.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]:
```suggestion
'''Expensive special ship that is purporse-built around a single core part using [[tech SHP_SPACE_FLUX_COMPOSITION]]. It has one core slot and one external slot, and its max [[metertype METER_STRUCTURE]] is high.
```

File: default/scripting/ship_parts/ShortRange/SR_FLUX_LANCE.focs.txt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+               [[COMBAT_TARGETS_VISIBLE_ENEMY]]
+               Or [
+                    Fighter
+                    [[COMBAT_TARGETS_NOT_DESTROYED_SHIP]]
+               ]
+           ]
+           And [
+               // Attack fighters starting from bout 2 until FIRST_COMBAT_ROUND_IN_CLOSE_TARGETING_RANGE
+               (CombatBout >= 2)
+               [[COMBAT_TARGETS_VISIBLE_ENEMY]] Fighter
```
Comment (geoffthemedio) [inline default/scripting/ship_parts/ShortRange/SR_FLUX_LANCE.focs.txt]:
```suggestion
               [[COMBAT_TARGETS_VISIBLE_ENEMY]]
               Fighter
```

**Comment by agrrr3** [comment]
addressed the review comments and did some further fixes on the descriptions; the hull descriptions were not up to date with the values.

**Comment by agrrr3** [comment]
Oh, and should I commit the gimp file i used to compose the flux-lance.png to the asset repository (maybe at art/icons/ship_parts/
?). Artistic depth is not high, i fiddled with the texture and colors in death_spore.blend and composed it on top of the death_spore.png (and added a color change layer)
As far as I can see the compositions of renders are not committed there currently.

**Comment by geoffthemedio** [comment]
If it's just overlaying existing available textures and tweaking colours, it's probably not super useful to archive the intermediate file format, but I'm not sure about that sort of thing.

File: default/stringtables/en.txt [hunk 7]
```diff
@@ -12057,6 +12063,7 @@ PLC_CHARGE_DESC
 • Weapon damage for ship weapons is increased by [[value PLC_CHARGE_DAMAGE_BOOST]] (i.e. for [[shippart SR_WEAPON_1_1]], [[shippart SR_WEAPON_2_1]], [[shippart SR_WEAPON_3_1]], or [[shippart SR_WEAPON_4_1]])
 • Number of fighters launched per [[FT_BAY_1]] is increased by [[value PLC_CHARGE_BAY_LAUNCH_BONUS]]
 • Number of shots for [[shippart SR_ARC_DISRUPTOR]] is increased by [[value PLC_CHARGE_ARC_DISRUPTOR_SHOT_BONUS]]
+• A charging [[shippart SR_FLUX_LANCE]] can attack ships a combat round earlier
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]:
```suggestion
• A charging [[shippart SR_FLUX_LANCE]] can attack ships starting one combat round earlier
```

File: default/stringtables/en.txt [hunk 8]
```diff
@@ -12729,17 +12736,29 @@ SHP_MIL_ROBO_CONT_DESC
...
+Ordinarily, propulsion in space is limited to either propellant or field propulsion. On relatively small scales however, it is possible to propel an object in space much like one could in water: by pushing backwards the medium in which the object is immersed. This requires the manipulation of space itself whereby the smallest units of space are collapsed, forcing the matter they once contained into adjacent units of space. This method of propulsion is very efficient, but dangerous on larger scales and easily detectable due to the dimensional effects. Further research into stealth technologies can reduce the dimensional disruption. The simplest solution to building a vessel using the flux propulsion is the bubble drive, which covers a sphere-shaped hull. More complex solutions are possible with better techs [[tech SHP_SPACE_FLUX_DRIVE]] and [[tech SHP_SPACE_FLUX_COMPOSITION]].
+
+SHP_SPACE_FLUX_COMPOSITION
+Spatial Flux Composition
+
+SHP_SPACE_FLUX_COMPOSITION_DESC
+By bending and stretching spatial flux bubbles and combining those with spatial flux drives one can build specialised hulls for specific purposes.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]:
```suggestion
By bending and stretching spatial flux bubbles and combining them with spatial flux drives, specialised hulls can be built for specific purposes.
```

File: default/stringtables/en.txt [hunk 9]
```diff
@@ -14903,6 +14922,18 @@ Spinal Antimatter Cannon
 SR_SPINAL_ANTIMATTER_DESC
 A huge spinal mount mass driver firing heavy projectiles made of antimatter. Attacks ships and planets.
 
+SR_FLUX_LANCE
+Spatial Flux Lance
+
+SR_FLUX_LANCE_DESC
+'''The Spatial Flux Lance, a internally mounted short-ranged ship weapon. Cheap to produce, but does only small amount of damage.
+
+The Spatial Flux Lance generator creates a short-lived spatial distortion in proximity outside the ship. The distortion is deadly to fighters and a nuisance for armoured ships. Multiple generators interfere with each other so only a single Flux Lance part may effectively mounted on a ship.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]:
```suggestion
The Spatial Flux Lance generator creates a short-lived spatial distortion in proximity to its ship. The distortion is deadly to fighters and a nuisance for armoured ships. Multiple generators interfere with each other, and thus only a single Flux Lance part may be mounted on each ship.
```

**Comment by Vezzra** [comment]
@agrrr3,

> Oh, and should I commit the gimp file i used to compose the flux-lance.png to the asset repository (maybe at art/icons/ship_parts/

Yes, please.

> As far as I can see the compositions of renders are not committed there currently.

There are a lot of images which don't have corresponding source files in the assets repo, because the assets repo was introduced only in 2014 (over a decade after the project started), and by that time we only had been able to recover the source files of a couple of the artwork in FO. Most of them we don't have.

Even after that we often/sometimes forgot to ask for the source files of artwork various people contributed, so unfortunately the assets repo is quite incomplete.

But that does not mean we still try to get every source file in we can... 😉

**Pull request merged**